src/xdplwf/recv.c

```diff
@@ -1367,157 +1367,177 @@ XdpGenericReceiveEnqueueTxNbl(
-                // first.
-                //
-                if (RxQueue->FragmentBufferInUse) {
-                    return;
-                }
+    FragmentExtension = XdpGetFragmentExtension(Frame, &RxQueue->FragmentExtension);
+    FragmentExtension->FragmentBufferCount = FragmentCount;
 
-                if (!XdpGenericReceiveLinearizeNb(RxQueue, *Nb)) {
-                    STAT_INC(&RxQueue->PcwStats, LinearizationFailures);
-                    goto Next;
-                }
+    //
+    // Store the original NB address so uninspected frames (e.g. those where
+    // virtual mappings failed) can be identified and dropped later.
+    //
+    InterfaceExtension =
+        XdpGetFrameInterfaceContextExtension(Frame, &RxQueue->FrameInterfaceContextExtension);
+    InterfaceExtension->Nb = Nb;
 
-                //
-                // The NB was converted to a single, virtually-contiguous buffer,
-                // so abandon fragmentation and proceed to XDP inspection.
-                //
-                FragmentCount = 0;
-                break;
-            }
+    //
+    // The NB has successfully been converted to XDP descriptors, so commit
+    // the descriptors to the XDP rings.
+    //
+    FragmentRing->ProducerIndex += FragmentCount;
+    FrameRing->ProducerIndex++;
 
-            //
-            // Reserve XDP descriptor space for this MDL. If space does not exist,
-            // flush existing descriptors to create space.
-            //
-            if (++FragmentCount > XdpRingFree(FragmentRing)) {
-                return;
-            }
+    return TRUE;
+}
 
-            Buffer =
-                XdpRingGetElement(
-                    FragmentRing,
-                    (FragmentRing->ProducerIndex + FragmentCount - 1) & FragmentRing->Mask);
-
-            Buffer->DataOffset = 0;
-            Buffer->DataLength = min(Mdl->ByteCount, DataLength);
-            Buffer->BufferLength = Mdl->ByteCount;
-            DataLength -= Buffer->DataLength;
-
-            SystemVa = XdpGetVirtualAddressExtension(Buffer, &RxQueue->BufferVaExtension);
-            SystemVa->VirtualAddress =
-                MmGetSystemAddressForMdlSafe(Mdl, LowPagePriority | MdlMappingNoExecute);
-            if (SystemVa->VirtualAddress == NULL || XdpLwfFaultInject()) {
-                STAT_INC(&RxQueue->PcwStats, MappingFailures);
-                goto Next;
-            }
-        }
+static
+BOOLEAN
+XdpGenericReceivePreInspectNbs(
+    _In_ XDP_LWF_GENERIC_RX_QUEUE *RxQueue,
+    _In_ BOOLEAN CanPend,
+    _Inout_ NET_BUFFER_LIST **Nbl,
+    _Inout_ NET_BUFFER **Nb
+    )
+{
+    BOOLEAN FlushNeeded = FALSE;
+    BOOLEAN InspectionNeeded = FALSE;
 
-        FragmentExtension = XdpGetFragmentExtension(Frame, &RxQueue->FragmentExtension);
-        FragmentExtension->FragmentBufferCount = FragmentCount;
+    //
+    // For low resources indications (`CanPend == TRUE`), ensure only one NBL is processed at a time
```

Comment:
1. low resources is iff `CanPend == FALSE`
2. we don't use markdown in comments, though that's an interesting idea

Reply:
Fixed.
Lots of tools now support markdown in comments for preview or documentation generation, and I tend to find it more readable as a human, especially in bigger comments (inline documentation kind of things)